### PR TITLE
Issue 71

### PR DESCRIPTION
**Pull request by venuiyengar**
Please review and merge this to dev branch.
This enables to run SQLXML from RDBMS, Transform to JSON and write the JSON contenet to file system specified by the output_file_path.

### CODE REVIEW

**Comment by sastafford** [comment]
@venuiyengar - How did you verify that this job works?  Do you have accompanying test code?

File: core/src/main/java/com/marklogic/spring/batch/item/SQLXMLtoJsonProcessor.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ */
+public class SQLXMLtoJsonProcessor extends LoggingObject
+implements ItemProcessor<Map<String, Object>, Map<String, Object>> {
+	public static final int PRETTY_PRINT_INDENT_FACTOR = 4;	
+    @Override
+    public Map<String, Object> process(Map<String, Object> columnMap) throws Exception {
+        Map<String, Object> newColumnMap = new LinkedHashMap<>();
+        for (String key : columnMap.keySet()) {
+            JSONObject xmlJSONObj = XML.toJSONObject(((SQLXML)columnMap.get(key)).getString());
```
Comment (sastafford) [inline core/src/main/java/com/marklogic/spring/batch/item/SQLXMLtoJsonProcessor.java]: Do you have any concern with using this XML to JSON transform library?  There is a risk of things getting lost in translation

**Comment by venuiyengar** [comment]
I tested by unzipping and running the job from command line and bat file. I do n’t have a test for this as it needs DB connection. May be I need to mock with some test.

File: core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * This writer writes the json file that has been extracted and processed from SQLXML to a defined location
+ * on the file system. The path of the file location is specified as part of the job configuration. The writer 
+ * constructs the filename based on the column name of the extracted xml from RDBMS.
+ * 
+ * @author viyengar
+ *
+ */
+public class JsonItemFileWriter extends AbstractDocumentWriter implements ItemWriter<Map<String, Object>> {
+
```
Comment (sastafford) [inline core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java]: A class that extends the AbstractDocumentWriter is intended to write to MarkLogic directly.  You are writing to the filesystem.  What about investigating an OOTB Spring Batch ItemWriter like MultiResourceItemWriter<T>?

File: core/src/main/java/com/marklogic/spring/batch/item/SQLXMLtoJsonProcessor.java (hunk 1 above)
Comment (venuiyengar) [inline core/src/main/java/com/marklogic/spring/batch/item/SQLXMLtoJsonProcessor.java]: As long as the library is available there is no issue. However, I have to do special casting as the data coming out from the DB was of object type and it needed this special cast for it to work.

File: core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java (hunk 2 above)
Comment (sastafford) [inline core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java]: @rjrudin  - For a class like a JsonItemFileWriter - should it be more specific for its template? 

```
public class JsonItemFileWriter implements ItemWriter<JSONObject>
```
We could force a string object, but we are losing any type of typing.  
```
public class JsonItemFileWriter implements ItemWriter<Map<String, String>>
```

[JSONObject](https://github.com/stleary/JSON-java/blob/master/JSONObject.java)

**Comment by rjrudin** [comment]
I think we have a small conundrum with writing tests for readers/writers.

We want resuable readers/writers to be in core, but I think we need to write tests for them in jobs, because we really want them to be part of a Config class, which goes into jobs. And we can leverage AbstractJobTest. And our test class in jobs can always make a temporary Config class to test out readers/writers.

File: core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java (hunk 2 above)
Comment (rjrudin) [inline core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java]: I agree with ItemWriter<JSONObject>.

We have to tread carefully here too. I think the org.json library is a good choice for converting XML to JSON. 

But do we want that to be part of core? We don't run core to include everything but the kitchen sink in terms of dependencies. 

I think Camel illustrates a good pattern. They have a separate project for every component. It's more overhead in terms of project management, but it's very nice for a user - I only specify the components I need, and I keep my classpath as simple as possible, thus avoiding classpath conflicts. 

So perhaps we want more than "core" and "jobs". Perhaps some of "jobs" goes into "core" - like anything that with making a Config class - and we divide "jobs" up into different "components", based on inputs and outputs. This reader/writer could be part of a SQLXML component, for instance.

**Comment by sastafford** [comment]
As of v04.1, I have the supporting test jobs in a separate jar now: [marklogic-spring-batch-test](https://bintray.com/sastafford/maven/marklogic-spring-batch-test/view).  This one is published to jcenter().  Any supporting test classes should go in there.

**Comment by sastafford** [comment]
Here is a [section](http://docs.spring.io/spring-batch/trunk/reference/html/testing.html) on Spring Batch unit testing. 

@venuiyengar - let me know your timelines.  If you need to get this stuff in and show the client something let us know and we can add an issue to add a test.

File: core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java (hunk 2 above)
Comment (sastafford) [inline core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java]: WRT the JSONObject, we need to align with engineering and use Jackson for JSON objects.  That's what is used in the Java Client API.  

@rjrudin - We are going to stick with core and jobs for now.  I think multiple components will complicate things.

Comment (rjrudin) [inline core/src/main/java/com/marklogic/spring/batch/item/JsonItemFileWriter.java]: SO does have an example of using Jackson for this - http://stackoverflow.com/questions/6746059/parsing-xml-into-json

@venuiyengar  can you try that sample?

Interestingly, I also found an SO thread on why json-lib (what @venuiyengar  is currently using) is a better choice than Jackson - http://stackoverflow.com/questions/15051973/converting-xml-to-json-using-jackson

So hopefully Jackson is up to the task of providing a reasonable transformation of XML to JSON, but if not, we'll probably need json-lib.

**Comment by sastafford** [comment]
@venuiyengar I'm going to close this pull request.  I'll keep the branch open for now, since I like the idea of using the SQLXML syntax.  But we need this pull request made against the DEV branch.

**Pull request closed without merging**